#!/usr/bin/env python

import os
import sys

from pytest.CondorJob import CondorJob
from pytest.CondorTest import CondorTest
from pytest.HTTPServer import HTTPServer
from pytest.PersonalCondor import PersonalCondor
from pytest.Utils import Utils


def SuccessCallback():
    Utils.TLog("Job successful as expected. Test passed.")
    sys.exit(0)

def FailureCallback():
    Utils.TLog("Job failed unexpectedly. Test failed.")
    sys.exit(1)

def HttpSuccessUrlHandler(request):
    request.protocol_version = "HTTP/1.1"
    request.send_response(200)
    request.send_header("Content-type", "text/html")
    request.end_headers()
    request.wfile.write("<html>Great success!</html>")


def main():

    # Create a new test environment
    test = CondorTest("cmd_curl_plugin_singlefile_success")

    # Start a HTTP server, and register the /success URL to return a 
    httpd = HTTPServer()
    httpd.Start()
    httpd.RegisterUrlHandler("/success", HttpSuccessUrlHandler)
    success_url = "http://127.0.0.1:" + str(httpd._port) + "/success"

    # Stand up a PersonalCondor
    personal_handle = test.StartPersonalCondor()
    if personal_handle is -1:
        Utils.TLog("Failed to start the PersonalCondor environment. Exiting.")
        sys.exit(1)
    else:
        Utils.TLog("PersonalCondor environment started successfully with handle " + str(personal_handle))

    # Setup a new test job
    job_args = {
        "executable":               "cmd_curl_plugin_singlefile_success.py",
        "output":                   "cmd_curl_plugin_singlefile_success.out",
        "error":                    "cmd_curl_plugin_singlefile_success.err",
        "transfer_input_files":     success_url,
        "should_transfer_files":    "YES"
    }
    test_job = CondorJob(job_args)

    # Register handlers
    test.RegisterSuccess(SuccessCallback)
    test.RegisterFailure(FailureCallback)

    # Submit the test job
    result = test.SubmitJob(test_job)
    Utils.TLog("Test returned result: " + str(result))

    # All done, sit and wait. Success/failure callbacks will exit the test.


if __name__ == "__main__":
    main()